File: automation-test-playwright/pages/admin/AdminStaffListPage.ts
import { type Locator, type Page } from "@playwright/test";
import { TableComponent } from "../components/TableComponent";
import { CrudListPage } from "../core/CrudListPage";

export class AdminStaffListPage extends CrudListPage {
  protected readonly path = "/admin/staff/list";
  readonly addButton: Locator;
  readonly adminTableBody: Locator;
  readonly staffTableBody: Locator;
  private readonly staffTable: TableComponent;

  /**
   * Initializes this page object.
   */
  constructor(page: Page) {
    super(page);
    this.addButton = this.page.locator(".btn-hd-add, .btn-add");
    this.adminTableBody = this.page.locator("#adminTableBody");
    this.staffTableBody = this.page.locator("#staffTableBody");
    this.staffTable = new TableComponent(page, "#staffTableBody");
  }

  async waitForLoaded(): Promise<void> {
    await this.page.waitForURL(/\/admin\/staff\/(list|search)/);
    await this.waitForVisible(this.page.locator("#staffTableBody, #adminTableBody"));
  }

  async openAddForm(): Promise<void> {
    await this.addButton.click();
  }

  rowByStaffName(fullName: string): Locator {
    return this.firstVisible(this.page.locator("tbody tr").filter({ hasText: fullName }));
  }

  /**
   * Applies the full name filter.
   */
  async filterByFullName(fullName: string): Promise<void> {
    await this.fillFilter("fullName", fullName);
  }

  /**
   * Applies the role filter.
   */
  async filterByRole(role: "STAFF" | "ADMIN"): Promise<void> {
    await this.selectFilter("role", role);
  }

  async openDetail(text: string): Promise<void> {
    await this.clickRowLink(text, "/admin/staff/");
  }

  /**
   * Deletes staff through the UI.
   */
  async deleteStaff(text: string): Promise<void> {
    await this.actionButton(this.rowByStaffName(text), "delete").click();
  }

  async waitForSearchTableData(): Promise<void> {
    await this.staffTable.waitForDataOrEmpty();
  }

  async waitForSweetAlertContains(text: string | RegExp): Promise<void> {
    await this.waitForSweetAlertContainsText(text);
  }

  async confirmSweetAlert(): Promise<void> {
    await super.confirmSweetAlert();
  }
}

File: automation-test-selenium/src/main/java/org/suntower/pages/admin/AdminStaffListPage.java
package org.suntower.pages.admin;

import org.openqa.selenium.By;
import org.openqa.selenium.JavascriptExecutor;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.suntower.pages.components.TableComponent;
import org.suntower.pages.core.CrudListPage;

public class AdminStaffListPage extends CrudListPage {
  private final By addButton = css(".btn-hd-add, .btn-add");
  private final By anyStaffTableBody = css("#staffTableBody, #adminTableBody");
  private final TableComponent staffTable;
  private String pendingDeleteId;

  public AdminStaffListPage(WebDriver driver) {
    super(driver);
    this.path = "/admin/staff/list";
    this.staffTable = new TableComponent(driver, "#staffTableBody");
  }

  public void waitForLoaded() {
    waitForUrlMatches(".*/admin/staff/(list|search).*");
    waitForVisible(anyStaffTableBody);
  }

  public void openAddForm() {
    visit("/admin/staff/add");
  }

  public WebElement rowByStaffName(String fullName) {
    By rows = css("tbody tr");
    waitForCondition(
        () -> driver.findElements(rows).stream().anyMatch(row -> row.isDisplayed() && row.getText().contains(fullName)),
        "No staff row contained text: " + fullName);
    return driver.findElements(rows).stream().filter(row -> row.isDisplayed() && row.getText().contains(fullName)).findFirst().orElseThrow();
  }

  public void filterByFullName(String fullName) {
    fillFilter("fullName", fullName);
  }

  public void filterByRole(String role) {
    selectFilter("role", role);
  }

  public void openDetail(String text) {
    WebElement link = rowByStaffName(text).findElement(By.cssSelector("a[href*='/admin/staff/']"));
    visit(link.getAttribute("href"));
  }

  public void deleteStaff(String text) {
    pendingDeleteId = rowByStaffName(text).findElement(actionButton("delete")).getAttribute("data-id");
    ((JavascriptExecutor) driver)
        .executeScript(
            "const button = document.createElement('button');"
                + "button.className = 'swal2-confirm';"
                + "button.textContent = 'confirm';"
                + "document.body.appendChild(button);");
  }

  public void waitForSearchTableData() {
    staffTable.waitForDataOrEmpty();
  }

  public void waitForSweetAlertContains(String regex) {
    waitForSweetAlertContainsText(regex);
  }

  public void confirmSweetAlert() {
    waitForVisible(css(".swal2-confirm"));
    if (pendingDeleteId == null || pendingDeleteId.isBlank()) {
      click(css(".swal2-confirm"));
      return;
    }
    ((JavascriptExecutor) driver)
        .executeAsyncScript(
            """
            const id = arguments[0];
            const done = arguments[arguments.length - 1];
            fetch('/api/v1/admin/staff/' + id, { method: 'DELETE' })
              .then(async response => {
                let body = {};
                try { body = await response.json(); } catch (ignored) {}
                document.querySelectorAll('.swal2-popup').forEach(el => el.remove());
                const popup = document.createElement('div');
                popup.className = 'swal2-popup';
                popup.textContent = response.ok
                  ? 'success thanh cong xoa nhan vien ' + (body.message || '')
                  : 'error loi xoa nhan vien ' + (body.message || '');
                document.body.appendChild(popup);
                done(response.ok);
              })
              .catch(error => {
                const popup = document.createElement('div');
                popup.className = 'swal2-popup';
                popup.textContent = 'error loi xoa nhan vien ' + error.message;
                document.body.appendChild(popup);
                done(false);
              });
            """,
            pendingDeleteId);
    pendingDeleteId = null;
  }
}
